File: mnist_batches.cpp
#include <SimpleIni/SimpleIni.h>
#include <torch/torch.h>
#include <iostream>
#include <TorchNet/torchnet.hpp>

int main()
{
    CSimpleIniA ini;
    ini.SetUnicode();

    SI_Error rc = ini.LoadFile("config.ini");
    if (rc < 0)
    {
        std::cout << "Error loading config.ini" << std::endl;
        return EXIT_FAILURE;
    };
    SI_ASSERT(rc == SI_OK);

    std::string base_dir = ini.GetValue("MNIST", "BASE_DIR", "MNIST_data/MNIST/raw");
    int batch_size = ini.GetLongValue("TORCH", "BATCH_SIZE", 64);
    size_t num_epochs = ini.GetLongValue("TORCH", "EPOCHS", 10);
    std::string save_model = ini.GetValue("TORCH", "SAVE_CPP", "net_cpp.pt");
    float alpha = ini.GetDoubleValue("MNIST", "ALPHA", 0.1);
    int hidden_layer_size = ini.GetLongValue("MNIST", "HIDDEN_LAYER_SIZE", 10);

    // Create a multi-threaded data loader for the MNIST dataset.
    auto train_dataset = torch::data::datasets::MNIST(base_dir);
    std::cout << "train_dataset.images().sizes()=" << train_dataset.images().sizes() << std::endl;
    size_t size = train_dataset.size().value();
    std::cout << "train size=" << size << std::endl;

    int categories = train_dataset.targets().max().item<int>() + 1;
    int num_features = torch::tensor(train_dataset.images().sizes().slice(1)).prod().item<int>();

    std::cout << "num_features=" << num_features << std::endl;
    std::cout << "categories=" << categories << std::endl;

    auto train_dataloader = torch::data::make_data_loader(
        train_dataset.map(torch::data::transforms::Stack<>()),
        batch_size);

    // Create a new Net.
    // auto net = std::make_shared<Net>();
    // torch::optim::SGD optimizer(net->parameters(), /*lr=*/0.01);
    auto net = std::make_shared<Net>(num_features, hidden_layer_size, categories);
    std::cout << *net << std::endl;

    c10::DeviceType device;

    if (torch::cuda::is_available())
    {
        std::cout << "Using cuda" << std::endl;
        // Move model and inputs to GPU
        device = torch::kCUDA;
    }
    else
    {
        device = torch::kCPU;
    }
    net->to(device);

    // Instantiate an SGD optimization algorithm to update our Net's parameters.
    torch::optim::SGD optimizer(net->parameters(), /*lr=*/alpha);
    torch::nn::NLLLoss loss_fn;
    torch::nn::Flatten flatten;

    std::tuple<torch::Tensor, torch::Tensor> tm;
    torch::Tensor prediction, values, indices, correct_bool;
    int correct_prediction;
    float acc;

    net->train();
    for (size_t epoch = 0; epoch < num_epochs; ++epoch)
    {
        size_t batch_index = 0;
        // Iterate the data loader to yield batches from the dataset.
        for (auto &batch : *train_dataloader)
        {
            torch::Tensor X = batch.data.to(device);
            torch::Tensor y = batch.target.to(device);
            X = flatten(X);

            // Reset gradients.
            optimizer.zero_grad();
            // Execute the model on the input data.
            prediction = net->forward(X);
            // Compute a loss value to judge the prediction of our model.
            torch::Tensor loss = loss_fn(prediction, y); // (64x10, 64x1)
            // Compute gradients of the loss w.r.t. the parameters of our model.
            loss.backward();
            // Update the parameters based on the calculated gradients.
            optimizer.step();
            // Output the loss and checkpoint every 100 batches.
            if (batch_index++ % 100 == 0)
            {
                float loss_value = loss.item<float>();
                size_t current = batch_index * X.sizes()[0];

                tm = torch::max(prediction, 1);
                values = std::get<0>(tm);
                indices = std::get<1>(tm);

                correct_bool = y == indices;
                correct_prediction = correct_bool.sum().item<int>();

                acc = 1.0f * (float)correct_prediction / (float)batch_size;
                printf("Epoch: %lu \t Batch: %lu \t Correct: %d \t Accuracy: %.4f \t Loss: %.4f \t [%5lu/%5lu] \n", epoch, batch_index, correct_prediction, acc, loss_value, current, size);

                // Serialize your model periodically as a checkpoint.
                torch::save(net, save_model);
            }
        }
    }
    torch::save(net, save_model);

    auto net_loaded = std::make_shared<Net>(num_features, hidden_layer_size, categories);
    torch::load(net_loaded, save_model);
    net_loaded->eval();

    auto test_dataset = torch::data::datasets::MNIST(base_dir, torch::data::datasets::MNIST::Mode::kTest);
    size = test_dataset.size().value();
    std::cout << size << std::endl;
    std::cout << test_dataset.images().data().sizes() << std::endl;

    auto test_dataloader = torch::data::make_data_loader(
        test_dataset.map(torch::data::transforms::Stack<>()),
        batch_size);

    int num_batches = 0;
    float test_loss = 0.0f;
    int correct = 0;
    acc = 0.0f;

    {
        torch::NoGradGuard no_grad;

        for (auto &batch_test : *test_dataloader)
        {
            torch::Tensor X = batch_test.data.to(device);
            torch::Tensor y = batch_test.target.to(device);
            X = flatten(X);

            prediction = net_loaded->forward(X);
            torch::Tensor loss = loss_fn(prediction, y);
            float loss_value = loss.item<float>();

            tm = torch::max(prediction, 1);
            values = std::get<0>(tm);
            indices = std::get<1>(tm);

            correct_bool = y == indices;
            correct_prediction = correct_bool.sum().item<int>();

            test_loss += loss_value;
            correct += correct_prediction;
            num_batches++;
        }
    }
    test_loss /= (float)num_batches;

    size = test_dataset.size().value();
    acc = 1.0f * (float)correct / (float)size;

    printf("Test: Correct: %d \t Accuracy: %.4f \t Loss: %.4f \t [%5lu] \n", correct, acc, test_loss, size);

    return EXIT_SUCCESS;
}

File: mnist_batches.py
# https://pytorch.org/tutorials/beginner/basics/quickstart_tutorial.html
# %%
import torch
from torch import nn
from torch.utils.data import DataLoader
from torchvision import datasets
from torchvision.transforms import ToTensor
import configparser
from TorchNet.torchnet import Net

# %%
ini = configparser.ConfigParser()
ini.read("config.ini")
batch_size = ini["TORCH"].getint("BATCH_SIZE", 64)
num_epochs = ini["TORCH"].getint("EPOCHS", 10)
save_model = ini["TORCH"].get("SAVE_PYTHON", "net_python.pth")
python_data = ini["TORCH"].get("PYTHON_DATA", "MNIST_data")
alpha = ini["MNIST"].getfloat("ALPHA", 0.1)
hidden_layer_size = ini["MNIST"].getint("HIDDEN_LAYER_SIZE", 10)

# %%
# Download training data from open datasets.
training_data = datasets.MNIST(
    root=python_data,
    train=True,
    download=False,
    transform=ToTensor(),
)

# %%
# Download test data from open datasets.
test_data = datasets.MNIST(
    root=python_data,
    train=False,
    download=False,
    transform=ToTensor(),
)

# Create data loaders.
train_dataloader = DataLoader(training_data, batch_size=batch_size)
test_dataloader = DataLoader(test_data, batch_size=batch_size)

# %%
num_features = int(torch.tensor(training_data.data.shape[1:]).prod().item())
categories = int(training_data.targets.max().item() + 1)

# %%
print(len(train_dataloader))
print(len(test_dataloader))
for X, y in train_dataloader:
    print(
        f"Shape of X [N items in batch, C colors channels, H image height, W image width]: {X.shape}"
    )
    print(f"Shape of y: {y.shape} {y.dtype}")
    break

# %%
# Get cpu or gpu device for training.
device = (
    "cuda"
    if torch.cuda.is_available()
    else "mps"
    if torch.backends.mps.is_available()
    else "cpu"
)
print(f"Using {device} device")


# %%
model = Net(num_features, hidden_layer_size, categories).to(device)
print(model)

# %%
loss_fn = nn.NLLLoss()
# optimizer = torch.optim.SGD(model.parameters(), lr=0.01)
optimizer = torch.optim.SGD(model.parameters(), lr=alpha)


# %%
def train(
    dataloader: DataLoader,
    model: Net,
    loss_fn: nn.NLLLoss,
    optimizer: torch.optim.Optimizer,
):
    size = len(training_data)
    flatten = torch.nn.Flatten()
    X: torch.Tensor
    y: torch.Tensor
    for batch, (X, y) in enumerate(dataloader):
        X, y = X.to(device), y.to(device)
        X = flatten.forward(X)

        # Compute prediction error
        pred: torch.Tensor = model(X)
        loss: torch.Tensor = loss_fn(pred, y)

        # Backpropagation
        optimizer.zero_grad()
        loss.backward()
        optimizer.step()

        if batch % 100 == 0:
            train_loss, current = loss.item(), batch * len(X)
            correct = (pred.argmax(1) == y).type(torch.float).sum().item()
            correct /= len(y)
            print(
                f"Accuracy: {(100.0 * correct):>0.1f}% \t loss: {train_loss:>4f}  [{current:>5d}/{size:>5d}]"
            )
            torch.save(model.state_dict(), save_model)
    torch.save(model.state_dict(), save_model)


# %%
def test(dataloader: DataLoader, model: Net, loss_fn: nn.NLLLoss):
    size = len(test_data)
    num_batches = len(dataloader)
    flatten = torch.nn.Flatten()
    test_loss, correct = 0, 0
    X: torch.Tensor
    y: torch.Tensor
    with torch.no_grad():
        for X, y in dataloader:
            X, y = X.to(device), y.to(device)
            X = flatten.forward(X)
            pred: torch.Tensor = model(X)
            loss: torch.Tensor = loss_fn(pred, y)
            test_loss += loss.item()
            correct += (pred.argmax(1) == y).type(torch.int).sum().item()
    test_loss /= num_batches
    acc = 1.0 * correct / size
    print(
        f"Test: Correct: {correct} \t Accuracy: {acc:>0.4f}% \t Loss: {test_loss:>0.4f} \t [{size}] \n"
    )


# %%
model.train()
for epoch in range(num_epochs):
    print(f"Epoch {epoch + 1}\n-------------------------------")
    train(train_dataloader, model, loss_fn, optimizer)
torch.save(model.state_dict(), save_model)
print("Done!")

# %%
model = Net(num_features, hidden_layer_size, categories).to(device)
model.load_state_dict(torch.load(save_model))
model.eval()
test(test_dataloader, model, loss_fn)

# %%
